From bdc938ac8353e9961a21b299e73d139cf2d6391c Mon Sep 17 00:00:00 2001
From: Mohammed SI ALI <mohammed.siali@softathome.com>
Date: Wed, 3 Sep 2025 16:42:51 +0200
Subject: [PATCH] Issue: PPW-975 fail to preclear dfs channels

Problem:
- cleareDfsChannels list only includes cleared channels of last
chanSwitch (zwdfs), never from pre-clearing
- acs never selects bw 160 of rad 5g chanspec

Cause:
- ath13 has redefined NL80211_CMD_STOP_BGRADAR_DETECT, leading to value
mismatch
- off-channel cac is taking much more time than fgdfs clearing, but pwhm
assumes bgdfs timeout with the shorter delay
- preclear requests for extension of cleared band (100/80->100/160)
which is rejects by nl80211, refusing to clear again already cleared
channels

Fix:
- fix NL80211_CMD_STOP_BGRADAR_DETECT definition
- define a bgclearTime per dfs channel, than can customized per vendor
- set default bgclear timeout in genPlugin to 7 times the fgdfs clearTime
retrieved from wiphy info: this prevents early expiration of zwdfs timer
- update setChanspec timer, with the duration of zwdfs timer based on
the clearing duration (fgcac or bgcac)
- when clearing a partially cleared band, select in genPlugin (for
nl80211 driver) the subPart not yet cleared, in order to prevent invalid
value error)

Closes PPW-975

Signed-off-by: Mohammed SI ALI <mohammed.siali@softathome.com>
---
 include/wld/wld.h                        |  3 +-
 include/wld/wld_channel.h                |  3 +
 include/wld/wld_nl80211_compat.h         |  8 ++-
 include/wld/wld_zwdfs.h                  |  1 +
 src/ChannelManagement/wld_chanmgt.c      | 13 ++--
 src/ChannelManagement/wld_channel.c      | 43 ++++++++++++-
 src/ChannelManagement/wld_channel_info.c | 53 +++++++++++++---
 src/ChannelManagement/wld_zwdfs.c        | 29 ++++++++-
 src/Plugin/wifiGen_rad.c                 | 80 ++++++++++++++++++++++--
 src/nl80211/wld_nl80211_api.c            |  1 +
 src/nl80211/wld_rad_nl80211.c            |  5 +-
 11 files changed, 209 insertions(+), 30 deletions(-)

diff --git a/include/wld/wld.h b/include/wld/wld.h
index b226963cf..f49bf9708 100644
--- a/include/wld/wld.h
+++ b/include/wld/wld.h
@@ -1399,7 +1399,8 @@ typedef struct {
     wld_channelChangeReason_e reason;
     char reasonExt[128];
     swl_timeMono_t changeTime;
-    swl_trl_e isApplied; //True:applied, False:failed, Unknown:ongoing
+    swl_trl_e isApplied;              //True:applied, False:failed, Unknown:ongoing
+    uint32_t estimatedChangeDuration; //estimated change duration in ms, including clear time when needed
 } wld_rad_detailedChanState_t;
 
 typedef enum {
diff --git a/include/wld/wld_channel.h b/include/wld/wld_channel.h
index 4207d7335..4fca6df42 100644
--- a/include/wld/wld_channel.h
+++ b/include/wld/wld_channel.h
@@ -97,6 +97,9 @@ int wld_channel_get_channel_to_clear(T_Radio* pRad, swl_bandwidth_e operatingCha
 void wld_channel_set_channel_clear_time(int channel, uint32_t time);
 int wld_channel_get_channel_clear_time(int channel);
 int wld_channel_get_band_clear_time(swl_chanspec_t chanspec);
+void wld_channel_set_channel_bg_clear_time(int channel, uint32_t time);
+int wld_channel_get_channel_bg_clear_time(int channel);
+int wld_channel_get_band_bg_clear_time(swl_chanspec_t chanspec);
 
 bool wld_channel_is_long_wait(swl_chanspec_t chanspec); //Deprecated
 bool wld_channel_is_long_wait_band(swl_chanspec_t chanspec);
diff --git a/include/wld/wld_nl80211_compat.h b/include/wld/wld_nl80211_compat.h
index 8b40bee80..52ba59789 100644
--- a/include/wld/wld_nl80211_compat.h
+++ b/include/wld/wld_nl80211_compat.h
@@ -98,7 +98,7 @@
 #define NL80211_SURVEY_INFO(a) VAL(a, NL80211_SURVEY_INFO_MAX, __NL80211_SURVEY_INFO_INVALID)
 
 #define IS_SPECIFIED(v, m, u) (((v) <= (m)) && ((v) > (u)))
-#define NL80211_CMD_IS_SPECIFIED(a) IS_SPECIFIED(c, NL80211_CMD_MAX, NL80211_CMD_UNSPEC)
+#define NL80211_CMD_IS_SPECIFIED(c) IS_SPECIFIED(c, NL80211_CMD_MAX, NL80211_CMD_UNSPEC)
 #define NL80211_ATTR_IS_SPECIFIED(a) IS_SPECIFIED(a, NL80211_ATTR_MAX, NL80211_ATTR_UNSPEC)
 #define NL80211_BAND_ATTR_IS_SPECIFIED(a) IS_SPECIFIED(a, NL80211_BAND_ATTR_MAX, __NL80211_BAND_ATTR_INVALID)
 #define NL80211_BAND_IFTYPE_ATTR_IS_SPECIFIED(a) IS_SPECIFIED(a, NL80211_BAND_IFTYPE_ATTR_MAX, __NL80211_BAND_IFTYPE_ATTR_INVALID)
@@ -307,7 +307,9 @@
 #define NL80211_ATTR_EML_CAPABILITY NL80211_ATTR(317)
 #define NL80211_ATTR_MLD_CAPA_AND_OPS NL80211_ATTR(318)
 
-//defined since kernel > 6.9
-#define NL80211_CMD_STOP_BGRADAR_DETECT ((NL80211_CMD_MAX >= 157) ? 157 : NL80211_CMD(155))
+//to be up-streamed, not yet in kernel 6.17
+#define NL80211_CMD_STOP_BGRADAR_DETECT \
+    (!NL80211_ATTR_IS_SPECIFIED(NL80211_ATTR_RADAR_BACKGROUND) ? NL80211_CMD_UNSPEC : \
+     ((NL80211_CMD_MAX >= 161) ? NL80211_CMD(158) : ((NL80211_CMD_MAX >= 157) ? NL80211_CMD(157) : NL80211_CMD(155))))
 
 #endif /* INCLUDE_WLD_WLD_NL80211_COMPAT_H_ */
diff --git a/include/wld/wld_zwdfs.h b/include/wld/wld_zwdfs.h
index 0aa80a68b..77a927a81 100644
--- a/include/wld/wld_zwdfs.h
+++ b/include/wld/wld_zwdfs.h
@@ -94,6 +94,7 @@ typedef struct {
     bool direct;
     amxp_timer_t* timer;
     swl_chanspec_t tgtChSpec;
+    uint32_t estimatedCacTime;
 } wld_zwdfs_fsmCtx_t;
 
 typedef struct {
diff --git a/src/ChannelManagement/wld_chanmgt.c b/src/ChannelManagement/wld_chanmgt.c
index f1fcefd57..069ee99b1 100644
--- a/src/ChannelManagement/wld_chanmgt.c
+++ b/src/ChannelManagement/wld_chanmgt.c
@@ -592,6 +592,12 @@ swl_rc_ne wld_chanmgt_setTargetChanspec(T_Radio* pR, swl_chanspec_t chanspec, bo
     pR->targetChanspec.reason = reason;
     pR->targetChanspec.changeTime = swl_time_getMonoSec();
     pR->targetChanspec.isApplied = SWL_TRL_UNKNOWN;
+    /* set timeout for the call
+     * if the band is cleared use default timeout
+     * if not, wait clear time in addition */
+    pR->targetChanspec.estimatedChangeDuration = WLD_CHANMGT_REQ_CHANGE_CS_TIMEOUT;
+    pR->targetChanspec.estimatedChangeDuration += wld_channel_is_band_passive(chanspec) ? wld_channel_get_band_clear_time(chanspec) : 0;
+
     pR->channelChangeReason = reason;
     if(reasonExt != NULL) {
         swl_str_copy(pR->targetChanspec.reasonExt, sizeof(pR->targetChanspec.reasonExt), reasonExt);
@@ -701,13 +707,8 @@ amxd_status_t _Radio_setChanspec(amxd_object_t* obj,
     /* register callId & cancel callback */
     swl_function_deferCb(&pR->callIdReqChanspec, func, ret, s_setChanspecCanceled, pR);
 
-    /* set timeout for the call
-     * if the band is cleared use default timeout
-     * if not, wait clear time in addition */
-    uint32_t timeout = WLD_CHANMGT_REQ_CHANGE_CS_TIMEOUT;
-    timeout += wld_channel_is_band_passive(chanspec) ? wld_channel_get_band_clear_time(chanspec) : 0;
     amxp_timer_new(&pR->timerReqChanspec, s_setChanspecTimeout, pR);
-    amxp_timer_start(pR->timerReqChanspec, timeout);
+    amxp_timer_start(pR->timerReqChanspec, pR->targetChanspec.estimatedChangeDuration);
 
     if(!direct) {
         /* schedule action */
diff --git a/src/ChannelManagement/wld_channel.c b/src/ChannelManagement/wld_channel.c
index 47094926a..6a9cdf2cd 100644
--- a/src/ChannelManagement/wld_channel.c
+++ b/src/ChannelManagement/wld_channel.c
@@ -120,7 +120,8 @@
 typedef struct {
     int channel;
     uint32_t flags;
-    uint32_t clearTime; // DFS clear time in milliseconds
+    uint32_t clearTime;   // FG DFS clear time in milliseconds
+    uint32_t bgClearTime; // BG DFS clear time in milliseconds
 } wld_channel_data;
 
 typedef struct {
@@ -300,6 +301,24 @@ static int s_chanGetClearTime(int channel, swl_freqBandExt_e freqBand) {
     return channel_info->clearTime;
 }
 
+/**
+ * set clear time for given channel if possible
+ */
+static void s_chanSetBgClearTime(int channel, swl_freqBandExt_e freqBand, uint32_t time) {
+    wld_channel_data* channel_info = get_channel_data(channel, freqBand);
+    ASSERTS_NOT_NULL(channel_info, , ME, "NULL");
+    channel_info->bgClearTime = time;
+}
+
+/**
+ * get clear time for given channel if possible
+ */
+static int s_chanGetBgClearTime(int channel, swl_freqBandExt_e freqBand) {
+    wld_channel_data* channel_info = get_channel_data(channel, freqBand);
+    ASSERTS_NOT_NULL(channel_info, 0, ME, "NULL");
+    return channel_info->bgClearTime;
+}
+
 //////////////////////////////////////////////////
 // External methods
 //////////////////////////////////////////////////
@@ -326,6 +345,7 @@ void wld_channel_init_channels(T_Radio* rad) {
         band->channels[i].channel = rad->possibleChannels[i];
         band->channels[i].flags = WLD_CHAN_INIT;
         band->channels[i].clearTime = 0;
+        band->channels[i].bgClearTime = 0;
     }
 }
 
@@ -613,6 +633,27 @@ int wld_channel_get_channel_clear_time(int channel) {
     }
 }
 
+/**
+ * Set the time in milliseconds that a channel should be cleared.
+ */
+void wld_channel_set_channel_bg_clear_time(int channel, uint32_t time) {
+    ASSERTS_TRUE(swl_channel_isDfs(channel), , ME, "%d not a DFS channel", channel);
+    s_chanSetBgClearTime(channel, SWL_FREQ_BAND_EXT_5GHZ, time);
+}
+
+/**
+ * Returns the time in milliseconds that a channel should be cleared.
+ */
+int wld_channel_get_channel_bg_clear_time(int channel) {
+    ASSERTS_TRUE(swl_channel_isDfs(channel), 0, ME, "%d not a DFS channel", channel);
+    uint32_t clearTime = s_chanGetBgClearTime(channel, SWL_FREQ_BAND_EXT_5GHZ);
+    if(clearTime > 0) {
+        return clearTime;
+    }
+    return s_chanGetClearTime(channel, SWL_FREQ_BAND_EXT_5GHZ);
+}
+
+
 bool wld_channel_areAdjacent(swl_chanspec_t chanspec1, swl_chanspec_t chanspec2) {
     swl_chanspec_t chanspec_low;
     swl_chanspec_t chanspec_high;
diff --git a/src/ChannelManagement/wld_channel_info.c b/src/ChannelManagement/wld_channel_info.c
index 89549bdc1..a9a58bbf7 100644
--- a/src/ChannelManagement/wld_channel_info.c
+++ b/src/ChannelManagement/wld_channel_info.c
@@ -143,25 +143,58 @@ int wld_channel_get_center_channel(swl_chanspec_t chanspec) {
     return swl_chanspec_getCentreChannel(&chanspec);
 }
 
+/**
+ * Run action for every channel of chanspec.
+ */
+typedef void (* chanActionCb_f)(swl_chanspec_t chanspec, swl_channel_t chan, void* data);
+
+static void s_runPerBandChan(swl_chanspec_t chanspec, chanActionCb_f fct, void* data) {
+    swl_channel_t temp_channel = swl_chanspec_getBaseChannel(&chanspec);
+    for(uint8_t i = 0; i < swl_chanspec_getNrChannelsInBand(&chanspec); i++) {
+        SWL_CALL(fct, chanspec, temp_channel, data);
+        temp_channel += CHANNEL_INCREMENT;
+    }
+}
+
 /**
  * Returns the time in milliseconds that it takes to clear this band.
  */
+static void s_getChanCacTime(swl_chanspec_t chanspec _UNUSED, swl_channel_t chan, void* data) {
+    ASSERTS_NOT_NULL(data, , ME, "NULL");
+    int* pTime = (int*) data;
+    int temp_time = wld_channel_get_channel_clear_time(chan);
+    if(temp_time > *pTime) {
+        *pTime = temp_time;
+    }
+}
+
 int wld_channel_get_band_clear_time(swl_chanspec_t chanspec) {
     if(chanspec.band != SWL_FREQ_BAND_EXT_5GHZ) {
         return 0;
     }
-    int nrChannels = swl_chanspec_getNrChannelsInBand(&chanspec);
     int time = 0;
-    int temp_time = 0;
-    int i = 0;
-    int temp_channel = swl_chanspec_getBaseChannel(&chanspec);
-    for(i = 0; i < nrChannels; i++) {
-        temp_time = wld_channel_get_channel_clear_time(temp_channel);
-        if(temp_time > time) {
-            time = temp_time;
-        }
-        temp_channel += CHANNEL_INCREMENT;
+    s_runPerBandChan(chanspec, s_getChanCacTime, &time);
+    return time;
+}
+
+/**
+ * Returns the time in milliseconds that it takes to clear this band, in background.
+ */
+static void s_getChanBgCacTime(swl_chanspec_t chanspec _UNUSED, swl_channel_t chan, void* data) {
+    ASSERTS_NOT_NULL(data, , ME, "NULL");
+    int* pTime = (int*) data;
+    int temp_time = wld_channel_get_channel_bg_clear_time(chan);
+    if(temp_time > *pTime) {
+        *pTime = temp_time;
     }
+}
+
+int wld_channel_get_band_bg_clear_time(swl_chanspec_t chanspec) {
+    if(chanspec.band != SWL_FREQ_BAND_EXT_5GHZ) {
+        return 0;
+    }
+    int time = 0;
+    s_runPerBandChan(chanspec, s_getChanBgCacTime, &time);
     return time;
 }
 
diff --git a/src/ChannelManagement/wld_zwdfs.c b/src/ChannelManagement/wld_zwdfs.c
index f5b80039e..7dc6c07bd 100644
--- a/src/ChannelManagement/wld_zwdfs.c
+++ b/src/ChannelManagement/wld_zwdfs.c
@@ -90,8 +90,9 @@ static wld_zwdfs_fsmState_e s_handleFsmStartEvent(const wld_zwdfs_fsmCtx_t* pCtx
         return ZWDFS_FSM_STATE_FG_CLEARING;
     }
     if(!wld_rad_is_5ghz(pRad) ||
+       !swl_chanspec_isDfs(pRad->targetChanspec.chanspec) ||
        !wld_channel_is_band_passive(pRad->targetChanspec.chanspec) ||
-       !isBgDfsEnabled) {
+       (pRad->bgdfs_config.status != BGDFS_STATUS_IDLE)) {
         SAH_TRACEZ_INFO(ME, "%s: ZW_DFS direct switch to %u", pRad->Name, pRad->targetChanspec.chanspec.channel);
         SWL_CALL(pRad->pFA->mfn_wrad_setChanspec, pRad, pCtx->direct);
         /* next state */
@@ -217,6 +218,27 @@ static void s_radioStatusChange(wld_radio_status_change_event_t* event) {
     case CM_RAD_FG_CAC:
     case CM_RAD_BG_CAC:
     case CM_RAD_BG_CAC_EXT:
+        if(isSameTgtChspec) {
+            if((pRad->detailedState) == CM_RAD_FG_CAC) {
+                fsm->fsmCtx.estimatedCacTime = wld_channel_get_band_clear_time(fsm->fsmCtx.tgtChSpec);
+            } else {
+                fsm->fsmCtx.estimatedCacTime = wld_channel_get_band_bg_clear_time(fsm->fsmCtx.tgtChSpec);
+            }
+            uint32_t zwdfsTmOut = WLD_ZWDFS_REQ_CHANGE_CS_TIMEOUT + fsm->fsmCtx.estimatedCacTime;
+            if(amxp_timer_remaining_time(fsm->fsmCtx.timer) > 0) {
+                amxp_timer_start(fsm->fsmCtx.timer, zwdfsTmOut);
+            }
+            uint32_t setChsTmOut = amxp_timer_remaining_time(pRad->timerReqChanspec);
+            if(zwdfsTmOut > setChsTmOut) {
+                SAH_TRACEZ_INFO(ME, "%s: update estimated setChanspec duration %d => %d",
+                                pRad->Name, setChsTmOut, zwdfsTmOut);
+                pRad->targetChanspec.estimatedChangeDuration = zwdfsTmOut;
+                if(setChsTmOut > 0) {
+                    amxp_timer_start(pRad->timerReqChanspec, zwdfsTmOut);
+                }
+            }
+        }
+        break;
     case CM_RAD_BG_CAC_NS:
     case CM_RAD_BG_CAC_EXT_NS:
         break;
@@ -274,11 +296,12 @@ swl_rc_ne wld_zwdfs_start(T_Radio* pRad, bool direct) {
     int ret = amxp_timer_new(&fsm->fsmCtx.timer, s_zwdfsTimeout, pRad);
     ASSERT_FALSE(ret != 0, SWL_RC_ERROR, ME, "%s: error timer", pRad->Name);
     int timeout = WLD_ZWDFS_REQ_CHANGE_CS_TIMEOUT;
-    timeout += wld_channel_get_band_clear_time(pRad->targetChanspec.chanspec);
+    swl_rc_ne rc = s_execFsm(fsm, ZWDFS_FSM_EVENT_START, &fsm->fsmCtx);
+    timeout += fsm->fsmCtx.estimatedCacTime;
     amxp_timer_start(fsm->fsmCtx.timer, timeout);
     SAH_TRACEZ_INFO(ME, "%s: start ZWDFS on %s timeout=%d", pRad->Name,
                     swl_typeChanspec_toBuf32(pRad->targetChanspec.chanspec).buf, timeout);
-    return s_execFsm(fsm, ZWDFS_FSM_EVENT_START, &fsm->fsmCtx);
+    return rc;
 }
 
 swl_rc_ne wld_zwdfs_stop(T_Radio* pRad) {
diff --git a/src/Plugin/wifiGen_rad.c b/src/Plugin/wifiGen_rad.c
index 937118ed2..ff1b6271f 100644
--- a/src/Plugin/wifiGen_rad.c
+++ b/src/Plugin/wifiGen_rad.c
@@ -66,6 +66,7 @@
 #include "wld/wld_radio.h"
 #include "wld/wld_channel.h"
 #include "wld/wld_chanmgt.h"
+#include "wld/wld_chaninfo.h"
 #include "wld/wld_util.h"
 #include "wld/wld_linuxIfUtils.h"
 #include "wld/wld_linuxIfStats.h"
@@ -447,6 +448,16 @@ static void s_updateBandAndStandard(T_Radio* pRad, wld_nl80211_bandDef_t bands[]
     pRad->supportedDataTransmitRates = pOperBand->supportedDataTransmitRates;
 }
 
+/*
+ * ETSI EN 301 893 Table D.1: DFS requirement values
+ * the duration of Off-Channel CAC performed in an ETSI regulatory domain must be longer than the
+ * on-channel CAC performed.
+ * These are defaults increasing Off-channel CAC duration:
+ * - On regular DFS channels, the CAC duration is increased from 1 minute to minimum 6 minutes.
+ * - On DFS weather radar channels, the CAC duration is increased from 10 minutes to minimum 60 minutes.
+ */
+#define BG_CAC_DUR_MULTIPLIER 7 // 6 + 1 margin
+
 void s_readChanInfo(T_Radio* pRad, wld_nl80211_bandDef_t* pOperBand) {
     ASSERT_NOT_NULL(pRad, , ME, "NULL");
     wld_channel_clear_flags(pRad);
@@ -459,9 +470,29 @@ void s_readChanInfo(T_Radio* pRad, wld_nl80211_bandDef_t* pOperBand) {
         }
         wld_channel_mark_available_channel(chanSpec);
         if(pChan->isDfs) {
-            SAH_TRACEZ_INFO(ME, "%s: mark radar req, channel [%d] clear_time [%d]", pRad->Name, chanSpec.channel, pChan->dfsCacTime);
             wld_channel_mark_radar_req_channel(chanSpec);
             wld_channel_set_channel_clear_time(chanSpec.channel, pChan->dfsCacTime);
+            uint32_t dfsCacTime = wld_channel_get_channel_clear_time(chanSpec.channel);
+
+            /* fix low values of ETSI dfs cac time for weather channels */
+            if(swl_channel_isWeather(chanSpec.channel) &&
+               (getCountryZone(pRad->regulatoryDomainIdx) == SWL_OP_CLASS_COUNTRY_EU) &&
+               (dfsCacTime < WLD_CHAN_DFS_EXTENDED_CLEAR_TIME_MS)) {
+                dfsCacTime = WLD_CHAN_DFS_EXTENDED_CLEAR_TIME_MS;
+                wld_channel_set_channel_clear_time(chanSpec.channel, WLD_CHAN_DFS_EXTENDED_CLEAR_TIME_MS);
+            }
+
+            /*
+             * calculate bg dfs cac time: assuming off-channel clearing by jumps.
+             * This bgcac time hw dependent,
+             * as the vendor may use a dedicated chain to do onchannel cac for the bgdfs.
+             * Therefore, the whm vendor module may adjust channel's bgdfs cac time
+             */
+            uint32_t bgdfsCacTime = dfsCacTime * BG_CAC_DUR_MULTIPLIER;
+
+            SAH_TRACEZ_INFO(ME, "%s: mark radar req, channel [%d] clear_time [%d] (hwVal:%d) bg_clear_time [%d]",
+                            pRad->Name, chanSpec.channel, dfsCacTime, pChan->dfsCacTime, bgdfsCacTime);
+            wld_channel_set_channel_bg_clear_time(chanSpec.channel, bgdfsCacTime);
         }
         switch(pChan->status) {
         case WLD_NL80211_CHAN_AVAILABLE: wld_channel_clear_passive_channel(chanSpec); break;
@@ -892,7 +923,7 @@ static swl_rc_ne s_checkAndStartZwDfs(T_Radio* pRad, bool direct) {
        !swl_channel_isDfs(pRad->targetChanspec.chanspec.channel) ||
        !wld_rad_isUpExt(pRad) ||
        (wld_chanmgt_getCurBw(pRad) > pRad->maxChannelBandwidth) ||
-       (pRad->bgdfs_config.status == BGDFS_STATUS_OFF)) {
+       (pRad->bgdfs_config.status != BGDFS_STATUS_IDLE)) {
         return SWL_RC_DONE;
     }
     if(wld_channel_is_band_passive(pRad->targetChanspec.chanspec)) {
@@ -1154,14 +1185,48 @@ swl_rc_ne wifiGen_rad_getSpectrumInfo(T_Radio* rad, bool update, amxc_llist_t* l
     return rc;
 }
 
+
+/*
+ * check whether full or half chanspec has to be cleared
+ * as nl80211 rejects bgdfs of already cleared channels
+ */
+static swl_rc_ne s_getChspecToBeCleared(swl_chanspec_t inCs, swl_chanspec_t* pOutCs) {
+    if(!swl_chanspec_isDfs(inCs) || !wld_channel_is_band_passive(inCs)) {
+        return SWL_RC_INVALID_PARAM;
+    }
+    swl_chanspec_t outCs = inCs;
+    if(inCs.bandwidth > SWL_BW_20MHZ) {
+        swl_chanspec_t h1 = SWL_CHANSPEC_NEW(inCs.channel, inCs.bandwidth - 1, inCs.band);
+        swl_chanspec_t h2 = SWL_CHANSPEC_NEW(swl_channel_getComplementaryBaseChannel(&inCs), inCs.bandwidth - 1, inCs.band);
+        bool isH1Passive = wld_channel_is_band_passive(h1);
+        bool isH2Passive = wld_channel_is_band_passive(h2);
+        if(isH1Passive && !isH2Passive) {
+            outCs = h1;
+        } else if(!isH1Passive && isH2Passive) {
+            outCs = h2;
+        }
+    }
+    W_SWL_SETPTR(pOutCs, outCs);
+    return SWL_RC_OK;
+}
+
 swl_rc_ne wifiGen_rad_bgDfsStartExt(T_Radio* pRad, wld_startBgdfsArgs_t* args) {
     ASSERT_NOT_NULL(pRad, SWL_RC_INVALID_PARAM, ME, "NULL");
     ASSERT_NOT_NULL(args, SWL_RC_INVALID_PARAM, ME, "NULL");
     ASSERT_TRUE(pRad->pFA->mfn_misc_has_support(pRad, NULL, "RADAR_BACKGROUND", 0), SWL_RC_ERROR, ME,
                 "%s: radar background not supported", pRad->Name);
     ASSERT_TRUE((wld_rad_isUpExt(pRad) && !wld_bgdfs_isRunning(pRad)), SWL_RC_ERROR, ME, "%s: not ready", pRad->Name);
-
-    swl_rc_ne rc = wld_rad_nl80211_bgDfsStart(pRad, args);
+    swl_chanspec_t inCs = SWL_CHANSPEC_NEW(args->channel, args->bandwidth ? : wld_chanmgt_getDefaultSupportedBandwidth(pRad), pRad->operatingFrequencyBand);
+    swl_rc_ne rc = s_getChspecToBeCleared(inCs, &inCs);
+    ASSERT_TRUE(swl_rc_isOk(rc), rc, ME, "%s: %s can not be cleared", pRad->Name, swl_typeChanspecExt_toBuf32Ref(&inCs).buf);
+    wld_startBgdfsArgs_t tmpArgs = *args;
+    tmpArgs.channel = inCs.channel;
+    tmpArgs.bandwidth = inCs.bandwidth;
+    SAH_TRACEZ_INFO(ME, "%s: starting bgDfs of %s (args:c:%d,b:%d) ",
+                    pRad->Name, swl_typeChanspecExt_toBuf32Ref(&inCs).buf,
+                    args->channel, args->bandwidth);
+
+    rc = wld_rad_nl80211_bgDfsStart(pRad, &tmpArgs);
     ASSERT_TRUE(swl_rc_isOk(rc), rc, ME, "%s: fail to start bgDfs", pRad->Name);
     // let bgdfs start event update the datamodel
     return rc;
@@ -1172,7 +1237,12 @@ swl_rc_ne wifiGen_rad_bgDfsStop(T_Radio* pRad) {
     ASSERTI_TRUE(wld_bgdfs_isRunning(pRad), SWL_RC_OK, ME, "%s: bgDfs not running", pRad->Name);
 
     swl_rc_ne rc = wld_rad_nl80211_bgDfsStop(pRad);
-    ASSERT_TRUE(swl_rc_isOk(rc), rc, ME, "%s: fail to stop bgDfs", pRad->Name);
+    if(!swl_rc_isOk(rc)) {
+        SAH_TRACEZ_ERROR(ME, "%s: fail to stop bgDfs", pRad->Name);
+        if(wld_rad_isUpExt(pRad)) {
+            return rc;
+        }
+    }
     wld_bgdfs_notifyClearEnded(pRad, DFS_RESULT_OTHER);
     wld_rad_updateState(pRad, false);
     return rc;
diff --git a/src/nl80211/wld_nl80211_api.c b/src/nl80211/wld_nl80211_api.c
index 8f73b7c5b..f453c49b3 100644
--- a/src/nl80211/wld_nl80211_api.c
+++ b/src/nl80211/wld_nl80211_api.c
@@ -1054,6 +1054,7 @@ swl_rc_ne wld_nl80211_bgDfsStart(wld_nl80211_state_t* state, uint32_t ifIndex, i
 swl_rc_ne wld_nl80211_bgDfsStop(wld_nl80211_state_t* state, uint32_t ifIndex, int8_t ifMloLinkId) {
     swl_rc_ne rc = SWL_RC_INVALID_PARAM;
     ASSERT_NOT_NULL(state, rc, ME, "NULL");
+    ASSERTI_TRUE(NL80211_CMD_IS_SPECIFIED(NL80211_CMD_STOP_BGRADAR_DETECT), SWL_RC_ERROR, ME, "cmd STOP_BGRADAR_DETECT not supported");
 
     SAH_TRACEZ_INFO(ME, "Stopping BG_DFS ifIndex(%d)", ifIndex);
 
diff --git a/src/nl80211/wld_rad_nl80211.c b/src/nl80211/wld_rad_nl80211.c
index f0226a296..97ea91c29 100644
--- a/src/nl80211/wld_rad_nl80211.c
+++ b/src/nl80211/wld_rad_nl80211.c
@@ -521,7 +521,10 @@ swl_rc_ne wld_rad_nl80211_bgDfsStart(T_Radio* pRadio, wld_startBgdfsArgs_t* args
     int32_t index = 0;
     wld_rad_nl80211_getFirstEnabledVapLinkInfo(pRadio, &index, &ifMloLinkId);
 
-    return wld_nl80211_bgDfsStart(wld_nl80211_getSharedState(), index, ifMloLinkId, bgDfsChanspec);
+    swl_rc_ne rc = wld_nl80211_bgDfsStart(wld_nl80211_getSharedState(), index, ifMloLinkId, bgDfsChanspec);
+    ASSERT_TRUE(swl_rc_isOk(rc), rc, ME, "%s: fail to start bgDfs (index:%d, ifMloLinkId:%d, chSpec:%s)",
+                pRadio->Name, index, ifMloLinkId, swl_typeChanspecExt_toBuf32Ref(&bgDfsChanspec).buf);
+    return rc;
 }
 
 swl_rc_ne wld_rad_nl80211_bgDfsStop(T_Radio* pRadio) {
-- 
GitLab

